File: scala/random-search/src/randomsearch/RandomSearch.scala
package randomsearch

import scala.util.Random._
import Numeric.Implicits._

object RandomSearch extends App {
  
  def randomVec[A : Numeric](low: A, high: A, n: Int, rand: => A) = {
    Seq.fill(n) { low + (high - low) * rand }
  }
  
  def cost[A : Numeric, B](as: Seq[A]) = {
    val z = implicitly[Numeric[A]].zero
    Option(as.foldLeft(z) { (a, b) => a + (b * b) })
  }
  
  def search[A : Numeric, B](randSpace: => Seq[A], cost: Seq[A] => Option[B], iterations: Int)(implicit cmp: Ordering[B]) = {
    
    def iter(best: (Option[B], Seq[A]), i: Int): (Option[B], Seq[A]) = {
      if (i == iterations) best
      else {
        val currentSpace = randSpace
        val currentCost = cost(currentSpace)
        
        val soln = best._1 match {
          case Some(prevCost) if (cmp.gt(currentCost.get, prevCost)) => best
          case _ => (currentCost, currentSpace)
        }
        
        iter(soln, i+1)
      }
    }
    
    iter((None, Nil), 0)
  }
  
  val solution = search[Float, Float](randomVec(-5.0f, 5.0f, 2, nextFloat), cost, 100)
  println("Score: " + solution._1.get + " Solution: " + solution._2)
}

File: julia/random-search/randomsearch.jl
random_vec(size, min, max) = map((x) -> min + (max - min) * x, rand(size))

sum_squares(itr) = reduce((v, elem) -> v + elem ^ 2, 0, itr)

function search(search_space, cost, num_itr)
	best = search_space()
	score = cost(best)
	for i = 1:(num_itr - 1)
		currSpace = search_space()
		currScore = cost(currSpace)
		if currScore < score
			best = currSpace
			score = currScore
		end
	end
	return score, best
end

search_space() = random_vec(2, -5.0, 5.0)
score, best = search(search_space, sum_squares, 100)
println("Score: " * string(score) * ", Solution: " * string(best))